{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "library(stringr)\n",
    "\n",
    "lapply(c(\"dplyr\",\"Seurat\",\"HGNChelper\",\"openxlsx\"), library, character.only = T)\n",
    "\n",
    "load('../data/reference.RData')\n",
    "\n",
    "sample_meta<-'AD_HS_00001\tMolecular characterization of selectively vulnerable neurons in Alzheimerâ€™s Disease\t20\tH.Sapiens\tAD\tSuperior frontal gyrus;Superior frontal gyrus;Superior frontal gyrus;Superior frontal gyrus;Superior frontal gyrus;Superior frontal gyrus;Superior frontal gyrus;Superior frontal gyrus;Superior frontal gyrus;Superior frontal gyrus;Entorhinal cortex;Entorhinal cortex;Entorhinal cortex;Entorhinal cortex;Entorhinal cortex;Entorhinal cortex;Entorhinal cortex;Entorhinal cortex;Entorhinal cortex;Entorhinal cortex\tBraak:0;0;0;2;2;2;2;6;6;6;0;0;0;2;2;2;2;6;6;6\tExcitatoryNeuron;InhibitoryNeuron;Oligodendrocyte;OPC;Astrocyte;Microglia;Endothelia;Pericyte\tGSE147528\tSingle-nucleus RNA-seq\t10x Genomics\tMale\t60;50;71;72;87;91;77;82;72;82;60;50;71;72;87;91;77;82;72;82\tSRR11422700;SRR11422701;SRR11422702;SRR11422703;SRR11422704;SRR11422705;SRR11422706;SRR11422707;SRR11422708;SRR11422709;SRR11422710;SRR11422711;SRR11422712;SRR11422713;SRR11422714;SRR11422715;SRR11422716;SRR11422717;SRR11422718;SRR11422719'\n",
    "split_meta<-unlist(strsplit(sample_meta, '\\t'))\n",
    "samples<-unlist(strsplit(split_meta[14],';'))\n",
    "disease<-rep(split_meta[5],length(samples))\n",
    "stage<-unlist(strsplit(unlist(strsplit(split_meta[7],':'))[2],';'))\n",
    "gender<-rep(split_meta[12],length(samples))\n",
    "age<-unlist(strsplit(split_meta[13],';'))\n",
    "\n",
    "sample_meta<-data.frame(samples=samples,disease=disease,stage=stage,gender=gender,age=age)\n",
    "sample_meta[sample_meta$stage=='0','disease']<-'Control'\n",
    "rownames(sample_meta)<-sample_meta$samples\n",
    "\n",
    "# args <- commandArgs(trailingOnly=TRUE)\n",
    "args<-c('../data/3/scte','../data/3/cell_umap.txt','SRR11422700','/home/wdeng3/workspace/software/sc-type','../../universal_data/rmsk/rmsk_GRCh38.txt')\n",
    "data_path<-args[1]\n",
    "out_path<-args[2]\n",
    "sample_tag<-args[3]\n",
    "sctype_path<-args[4]\n",
    "rmsk_path<-args[5]\n",
    "rmsk<-read.table(rmsk_path,sep='\\t',header=F)\n",
    "to_remove<-unlist(rmsk[!(rmsk$V12 %in% c('LINE','SINE','LTR')),'V11'])\n",
    "to_remove<-str_replace_all(to_remove,'-','\\\\.')\n",
    "\n",
    "reps<-unlist(rmsk[rmsk$V12 %in% c('LINE','SINE','LTR'),'V11'])\n",
    "reps<-str_replace_all(reps,'-','\\\\.')\n",
    "\n",
    "create_seurat<-function(sample_tag){\n",
    "    ## load data\n",
    "    scte.data <- t(read.csv(paste0(data_path,'/',sample_tag,'/',sample_tag,'.csv'),check.names=F, row.names = 1))\n",
    "    # remove repeats from gene expression matrix\n",
    "\n",
    "    scte.gene<-scte.data[!(row.names(scte.data) %in% to_remove),]\n",
    "    # normalize and scale data\n",
    "    scte <- CreateSeuratObject(counts = scte.gene, project = sample_tag, min.cells = 3, min.features = 200)\n",
    "    # normalize data\n",
    "    scte[[\"percent.mt\"]] <- PercentageFeatureSet(scte, pattern = \"^MT.|^MT-\")\n",
    "    scte <- subset(scte, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)\n",
    "    scte <- SCTransform(scte, vars.to.regress = \"percent.mt\", verbose = FALSE)\n",
    "\n",
    "    cell_meta<-scte@meta.data\n",
    "    cell_meta$disease<-sample_meta[cell_meta$orig.ident,'disease']\n",
    "\n",
    "    cell_meta$stage<-sample_meta[cell_meta$orig.ident,'stage']\n",
    "    cell_meta$gender<-sample_meta[cell_meta$orig.ident,'gender']\n",
    "    cell_meta$age<-sample_meta[cell_meta$orig.ident,'age']\n",
    "    scte@meta.data<-cell_meta\n",
    "    scte\n",
    "}\n",
    "\n",
    "scte<-create_seurat(sample_tag)\n",
    "\n",
    "scte <- FindVariableFeatures(scte, selection.method = \"vst\", nfeatures = 2000)\n",
    "# scale and run PCA\n",
    "scte <- ScaleData(scte, features = rownames(scte))\n",
    "scte <- RunPCA(scte, features = VariableFeatures(object = scte))\n",
    "\n",
    "sdv<-Stdev(scte,reduction='pca')\n",
    "sdv<-sdv[sdv>2]\n",
    "npca<-length(sdv)\n",
    "\n",
    "scte <- FindNeighbors(scte, dims = 1:npca)\n",
    "scte <- FindClusters(scte, resolution = 0.8)\n",
    "scte <- RunUMAP(scte, dims = 1:npca)\n",
    "\n",
    "\n",
    "scte.anchors <- FindTransferAnchors(reference = reference, query = scte,\n",
    "    dims = 1:30, reference.reduction = \"pca\")\n",
    "scte.query <- MapQuery(anchorset = scte.anchors, reference = reference, query = scte,\n",
    "    refdata = list(celltype = \"broad.cell.type\"), reference.reduction = \"pca\", reduction.model = \"umap\")\n",
    "cell_meta<-scte.query@meta.data\n",
    "\n",
    "cell_meta$UMAP_1<-scte.query@reductions[['umap']]@cell.embeddings[,1]\n",
    "cell_meta$UMAP_2<-scte.query@reductions[['umap']]@cell.embeddings[,2]\n",
    "scte.query@meta.data<-cell_meta\n",
    "write.table(cell_meta,out_path,sep='\\t',quote=F,row.names=T)\n",
    "\n",
    "## get gene expression from scte_query\n",
    "scte.query.gene<-as.data.frame(t(as.data.frame(GetAssayData(scte.query, slot = \"data\"))))\n",
    "scte.query.gene$UMAP_1<-scte.query@meta.data$UMAP_1\n",
    "scte.query.gene$UMAP_2<-scte.query@meta.data$UMAP_2\n",
    "write.table(scte.query.gene,'../data/3/cell_exp.txt',sep='\\t',quote=F,row.names=T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_307/3827847824.py:7: DtypeWarning: Columns (0,1,2,3,4,6,7,8,13,14,15,16) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  rmsk=pd.read_table('../../universal_data/rmsk/rmsk_GRCh38.txt',sep='\\t',header=None)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AD_HS_00003AD_HS_00006\n",
      "AD_HS_00002AD_HS_00005AD_HS_00008AD_HS_00004MS_HS_00001MS_HS_00002PD_HS_00001AD_HS_00001\n",
      "loading data: AD_HS_00003_112AD_HS_00007\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "loading data: AD_HS_00006_52\n",
      "\n",
      "loading data: AD_HS_00004_998loading data: AD_HS_00002_801loading data: AD_HS_00005_469loading data: AD_HS_00008_588loading data: MS_HS_00001_159loading data: PD_HS_00001_174loading data: MS_HS_00002_133loading data: AD_HS_00001_55\n",
      "\n",
      "\n",
      "loading data: AD_HS_00007_205\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "['AD_HS_00001', 'AD_HS_00002', 'AD_HS_00003', 'AD_HS_00004', 'AD_HS_00005', 'AD_HS_00006', 'AD_HS_00007', 'AD_HS_00008', 'MS_HS_00001', 'MS_HS_00002', 'PD_HS_00001']\n",
      "extracting disease v.s. control\n",
      "extracting cell v.s. cell\n",
      "extracting disease v.s. control\n",
      "extracting cell v.s. cell\n",
      "extracting disease v.s. control\n",
      "extracting cell v.s. cell\n",
      "extracting disease v.s. control\n",
      "extracting cell v.s. cell\n",
      "extracting disease v.s. control\n",
      "extracting disease v.s. control\n",
      "extracting cell v.s. cell\n",
      "extracting cell v.s. cell\n",
      "extracting disease v.s. control\n",
      "extracting cell v.s. cell\n",
      "extracting disease v.s. control\n",
      "extracting cell v.s. cell\n",
      "extracting disease v.s. control\n",
      "extracting cell v.s. cell\n",
      "extracting disease v.s. control\n",
      "extracting cell v.s. cell\n",
      "extracting disease v.s. control\n",
      "extracting cell v.s. cell\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.stats import mannwhitneyu\n",
    "import os\n",
    "\n",
    "file_list=[x for x in os.listdir('../data/all_datasets') if x.endswith('.cell_exp.txt')]\n",
    "rmsk=pd.read_table('../../universal_data/rmsk/rmsk_GRCh38.txt',sep='\\t',header=None)\n",
    "rtes=rmsk[rmsk[11].isin(['LINE','SINE','LTR'])][10].unique()\n",
    "datasets=list(set([x.split('.')[0] for x in file_list]))\n",
    "datasets.sort()\n",
    "cell_types=['Ex','Ast','Mic','OPC','Oli','In']\n",
    "import random\n",
    "def get_dataset(dataset):\n",
    "    print(dataset)\n",
    "    randint=random.randint(0,1000)\n",
    "    print(f'loading data: {dataset}_{randint}')\n",
    "    results=[]\n",
    "    dt_ls=[x for x in file_list if x.startswith(dataset)]\n",
    "    cell_exp=pd.read_table('../data/all_datasets/'+dt_ls[0],index_col=0)\n",
    "    if len(dt_ls) >1:\n",
    "        cell_umap=pd.read_table(f'../data/all_datasets/{dataset}.1.cell_umap.txt',index_col=0)\n",
    "        for i in range(1,len(dt_ls)):\n",
    "            cell_exp=pd.concat([cell_exp,pd.read_table('../data/all_datasets/'+dt_ls[i],index_col=0)])\n",
    "            cell_umap=pd.concat([cell_umap,pd.read_table(f'../data/all_datasets/{dataset}.{i}.cell_umap.txt',index_col=0)])\n",
    "    else:\n",
    "        cell_umap=pd.read_table('../data/all_datasets/'+dataset+'.cell_umap.txt',index_col=0)\n",
    "    \n",
    "    cell_umap['predicted.celltype'] = cell_umap['predicted.celltype'].replace(\n",
    "        'Opc', 'OPC')\n",
    "    for i in range(cell_umap.shape[0]):\n",
    "        if cell_umap.iloc[i,1] =='Stage_0':\n",
    "            cell_umap.iloc[i,1]='Control'\n",
    "        if cell_umap.iloc[i,1] !='Control':\n",
    "            cell_umap.iloc[i,1]=cell_umap.iloc[i,7].split('_')[0]\n",
    "\n",
    "    cell_exp=cell_exp.loc[:,[x for x in cell_exp.columns if x in rtes]]\n",
    "    ## disease v.s. control\n",
    "    print('extracting disease v.s. control')\n",
    "    disease=dataset.split('_')[0]\n",
    "    for cell in cell_types:\n",
    "        disease_list=cell_umap.loc[(cell_umap['predicted.celltype']==cell) & (cell_umap['Diagnosis']!='Control'),:].index\n",
    "        control_list=cell_umap.loc[(cell_umap['predicted.celltype']==cell) & (cell_umap['Diagnosis']=='Control'),:].index\n",
    "        for te in rtes:\n",
    "            if te in cell_exp.columns:\n",
    "                disease_exp=np.exp(cell_exp.loc[disease_list,te])\n",
    "                control_exp=np.exp(cell_exp.loc[control_list,te])\n",
    "                pval=mannwhitneyu(disease_exp,control_exp)[1]\n",
    "                c1_median=np.median(disease_exp)\n",
    "                c2_median=np.median(control_exp)\n",
    "                fc=np.log2(np.median(disease_exp)/np.median(control_exp))\n",
    "                results.append([te,f'{disease} v.s. Control',dataset,cell,'NA',np.log(c1_median),np.log(c2_median),fc,pval])\n",
    "    ## cell v.s. cell\n",
    "    print('extracting cell v.s. cell')\n",
    "    for cell1 in cell_types:\n",
    "        for cell2 in cell_types:\n",
    "            if not cell1==cell2:\n",
    "                for condition in [disease,'Control']:\n",
    "                    cell1_list=cell_umap.loc[(cell_umap['predicted.celltype']==cell1) & (cell_umap['Diagnosis']==condition),:].index\n",
    "                    cell2_list=cell_umap.loc[(cell_umap['predicted.celltype']==cell2) & (cell_umap['Diagnosis']==condition),:].index\n",
    "                    for te in rtes:\n",
    "                        if te in cell_exp.columns:\n",
    "                            cell1_exp=np.exp(cell_exp.loc[cell1_list,te])\n",
    "                            cell2_exp=np.exp(cell_exp.loc[cell2_list,te])\n",
    "                            pval=mannwhitneyu(cell1_exp,cell2_exp)[1]\n",
    "                            c1_median=np.median(cell1_exp)\n",
    "                            c2_median=np.median(cell2_exp)\n",
    "                            fc=np.log2(np.median(cell1_exp)/np.median(cell2_exp))\n",
    "                            results.append([te,f'{cell1} v.s. {cell2}',dataset,'NA',condition,np.log(c1_median),np.log(c2_median),fc,pval])\n",
    "    return results\n",
    "from multiprocessing import Pool\n",
    "pool = Pool(20)\n",
    "print(datasets)\n",
    "ret=pool.map(get_dataset,datasets)\n",
    "pool.close()\n",
    "pool.join()\n",
    "\n",
    "final_result=[]\n",
    "for i in ret:\n",
    "    final_result.extend(i)\n",
    "final_result=pd.DataFrame(final_result,columns=['TE','Comparison','Dataset','Cell','Condition','log2_median1','log2_median2','log2_FC','pval'])\n",
    "final_result.to_csv('../data/all_datasets/ret_compare.txt',sep='\\t',index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "432366"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "te_exp=open(out_path+'/te_exp_boxplot.sql','w')\n",
    "\n",
    "te_exp.write('''use scARE;\n",
    "    DROP TABLE IF EXISTS `TE_EXP_BOXPLOT`;\n",
    "    CREATE TABLE `TE_EXP_BOXPLOT` (\n",
    "        ID int(11) NOT NULL AUTO_INCREMENT,\n",
    "        CELL_TYPE varchar(255) NOT NULL,\n",
    "        CELL_NUM int(11) NOT NULL,\n",
    "        DATASET varchar(255) NOT NULL,\n",
    "        DISEASE varchar(255) NOT NULL,\n",
    "        TE varchar(255) NOT NULL,\n",
    "        MAX float NOT NULL,\n",
    "        MIN float NOT NULL,\n",
    "        Q1 float NOT NULL,\n",
    "        MEDIAN float NOT NULL,\n",
    "        Q3 float NOT NULL,\n",
    "        PRIMARY KEY (ID));\n",
    "        set autocommit=0;\\n''')\n",
    "\n",
    "result=pd.DataFrame(result,columns=['cell_type','cell_num','dataset','disease','te','max','min','q1','median','q3'])\n",
    "result['cell_type']=result['cell_type'].apply(lambda x:'\"'+x+'\"')\n",
    "result['dataset']=result['dataset'].apply(lambda x:'\"'+x+'\"')   \n",
    "result['disease']=result['disease'].apply(lambda x:'\"'+x+'\"')\n",
    "result['te']=result['te'].apply(lambda x:'\"'+x+'\"')\n",
    "result.fillna(-1,inplace=True)\n",
    "\n",
    "count=0\n",
    "values_list=[]\n",
    "template='INSERT INTO `TE_EXP_BOXPLOT` (CELL_TYPE,CELL_NUM,DATASET,DISEASE,TE,MAX,MIN,Q1,MEDIAN,Q3) VALUES {values};\\n'\n",
    "for i in range(result.shape[0]):\n",
    "    if count==2000:\n",
    "        te_exp.write(template.format(values=','.join(values_list)))\n",
    "        count=0\n",
    "        values_list=[]\n",
    "    values_list.append('('+','.join([str(x) for x in result.iloc[i,:]])+')')\n",
    "    count+=1\n",
    "\n",
    "if count>0:\n",
    "    te_exp.write(template.format(values=','.join(values_list)))\n",
    "te_exp.write('commit;')\n",
    "te_exp.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TE</th>\n",
       "      <th>Comparison</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Cell</th>\n",
       "      <th>Condition</th>\n",
       "      <th>log2_median1</th>\n",
       "      <th>log2_median2</th>\n",
       "      <th>log2_FC</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>L1P5</td>\n",
       "      <td>AD v.s. Control</td>\n",
       "      <td>AD_HS_00001</td>\n",
       "      <td>Ex</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.374865e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AluY</td>\n",
       "      <td>AD v.s. Control</td>\n",
       "      <td>AD_HS_00001</td>\n",
       "      <td>Ex</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.966901</td>\n",
       "      <td>5.996269</td>\n",
       "      <td>-0.042369</td>\n",
       "      <td>2.691087e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L1MB5</td>\n",
       "      <td>AD v.s. Control</td>\n",
       "      <td>AD_HS_00001</td>\n",
       "      <td>Ex</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.297829</td>\n",
       "      <td>2.315329</td>\n",
       "      <td>-0.025247</td>\n",
       "      <td>5.616159e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AluSc</td>\n",
       "      <td>AD v.s. Control</td>\n",
       "      <td>AD_HS_00001</td>\n",
       "      <td>Ex</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.666867</td>\n",
       "      <td>4.714000</td>\n",
       "      <td>-0.067998</td>\n",
       "      <td>3.252002e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HAL1</td>\n",
       "      <td>AD v.s. Control</td>\n",
       "      <td>AD_HS_00001</td>\n",
       "      <td>Ex</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.322802</td>\n",
       "      <td>3.349713</td>\n",
       "      <td>-0.038824</td>\n",
       "      <td>3.574014e-04</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      TE       Comparison      Dataset Cell Condition  log2_median1  \\\n",
       "0   L1P5  AD v.s. Control  AD_HS_00001   Ex        NA      0.000000   \n",
       "1   AluY  AD v.s. Control  AD_HS_00001   Ex        NA      5.966901   \n",
       "2  L1MB5  AD v.s. Control  AD_HS_00001   Ex        NA      2.297829   \n",
       "3  AluSc  AD v.s. Control  AD_HS_00001   Ex        NA      4.666867   \n",
       "4   HAL1  AD v.s. Control  AD_HS_00001   Ex        NA      3.322802   \n",
       "\n",
       "   log2_median2   log2_FC          pval  \n",
       "0      0.000000  0.000000  1.374865e-02  \n",
       "1      5.996269 -0.042369  2.691087e-11  \n",
       "2      2.315329 -0.025247  5.616159e-04  \n",
       "3      4.714000 -0.067998  3.252002e-14  \n",
       "4      3.349713 -0.038824  3.574014e-04  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_result.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index([], dtype='object')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "control_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>predicted.celltype</th>\n",
       "      <th>Diagnosis</th>\n",
       "      <th>stage</th>\n",
       "      <th>msex</th>\n",
       "      <th>age_death</th>\n",
       "      <th>UMAP_1</th>\n",
       "      <th>UMAP_2</th>\n",
       "      <th>dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR11422700_AAAAAACTCCAACCAA</th>\n",
       "      <td>Ex</td>\n",
       "      <td>Stage_0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.892934</td>\n",
       "      <td>6.910680</td>\n",
       "      <td>AD_HS_00003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR11422700_AAAACAACAATGGAAT</th>\n",
       "      <td>Ex</td>\n",
       "      <td>Stage_0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.549984</td>\n",
       "      <td>6.959525</td>\n",
       "      <td>AD_HS_00003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR11422700_AAAACTCAGACAAAGG</th>\n",
       "      <td>Ex</td>\n",
       "      <td>Stage_0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.793271</td>\n",
       "      <td>6.914708</td>\n",
       "      <td>AD_HS_00003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR11422700_AAAACTCAGCGTTTAC</th>\n",
       "      <td>Ex</td>\n",
       "      <td>Stage_0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.781046</td>\n",
       "      <td>6.445986</td>\n",
       "      <td>AD_HS_00003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR11422700_AAAACTCTCCAAACAC</th>\n",
       "      <td>Ex</td>\n",
       "      <td>Stage_0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.880208</td>\n",
       "      <td>7.098388</td>\n",
       "      <td>AD_HS_00003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR11422709_TTTGTCACATGAACCT</th>\n",
       "      <td>Oli</td>\n",
       "      <td>Stage_6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-6.770343</td>\n",
       "      <td>1.267352</td>\n",
       "      <td>AD_HS_00003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR11422709_TTTGTCAGTCACCTAA</th>\n",
       "      <td>Ex</td>\n",
       "      <td>Stage_6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.176651</td>\n",
       "      <td>4.918381</td>\n",
       "      <td>AD_HS_00003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR11422709_TTTGTCAGTGTGGTTT</th>\n",
       "      <td>Oli</td>\n",
       "      <td>Stage_6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-6.788431</td>\n",
       "      <td>1.532442</td>\n",
       "      <td>AD_HS_00003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR11422709_TTTGTCATCACGGTTA</th>\n",
       "      <td>In</td>\n",
       "      <td>Stage_6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.986920</td>\n",
       "      <td>0.141141</td>\n",
       "      <td>AD_HS_00003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR11422709_TTTGTCATCGCGGATC</th>\n",
       "      <td>Ex</td>\n",
       "      <td>Stage_6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.906748</td>\n",
       "      <td>5.158192</td>\n",
       "      <td>AD_HS_00003</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100872 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             predicted.celltype Diagnosis  stage  msex  \\\n",
       "SRR11422700_AAAAAACTCCAACCAA                 Ex   Stage_0    NaN   NaN   \n",
       "SRR11422700_AAAACAACAATGGAAT                 Ex   Stage_0    NaN   NaN   \n",
       "SRR11422700_AAAACTCAGACAAAGG                 Ex   Stage_0    NaN   NaN   \n",
       "SRR11422700_AAAACTCAGCGTTTAC                 Ex   Stage_0    NaN   NaN   \n",
       "SRR11422700_AAAACTCTCCAAACAC                 Ex   Stage_0    NaN   NaN   \n",
       "...                                         ...       ...    ...   ...   \n",
       "SRR11422709_TTTGTCACATGAACCT                Oli   Stage_6    NaN   NaN   \n",
       "SRR11422709_TTTGTCAGTCACCTAA                 Ex   Stage_6    NaN   NaN   \n",
       "SRR11422709_TTTGTCAGTGTGGTTT                Oli   Stage_6    NaN   NaN   \n",
       "SRR11422709_TTTGTCATCACGGTTA                 In   Stage_6    NaN   NaN   \n",
       "SRR11422709_TTTGTCATCGCGGATC                 Ex   Stage_6    NaN   NaN   \n",
       "\n",
       "                              age_death    UMAP_1    UMAP_2      dataset  \n",
       "SRR11422700_AAAAAACTCCAACCAA        NaN  3.892934  6.910680  AD_HS_00003  \n",
       "SRR11422700_AAAACAACAATGGAAT        NaN  4.549984  6.959525  AD_HS_00003  \n",
       "SRR11422700_AAAACTCAGACAAAGG        NaN  3.793271  6.914708  AD_HS_00003  \n",
       "SRR11422700_AAAACTCAGCGTTTAC        NaN  5.781046  6.445986  AD_HS_00003  \n",
       "SRR11422700_AAAACTCTCCAAACAC        NaN  7.880208  7.098388  AD_HS_00003  \n",
       "...                                 ...       ...       ...          ...  \n",
       "SRR11422709_TTTGTCACATGAACCT        NaN -6.770343  1.267352  AD_HS_00003  \n",
       "SRR11422709_TTTGTCAGTCACCTAA        NaN  1.176651  4.918381  AD_HS_00003  \n",
       "SRR11422709_TTTGTCAGTGTGGTTT        NaN -6.788431  1.532442  AD_HS_00003  \n",
       "SRR11422709_TTTGTCATCACGGTTA        NaN  3.986920  0.141141  AD_HS_00003  \n",
       "SRR11422709_TTTGTCATCGCGGATC        NaN  2.906748  5.158192  AD_HS_00003  \n",
       "\n",
       "[100872 rows x 8 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_umap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## assign cell type according to ScType tuturial\n",
    "\n",
    "## load sc type and database\n",
    "# source(paste0(sctype_path,\"/R/gene_sets_prepare.R\"))\n",
    "# source(paste0(sctype_path,\"/R/sctype_score_.R\"))\n",
    "# gs_list = gene_sets_prepare(paste0(sctype_path,'/ScTypeDB_full.xlsx'), \"Brain\" )\n",
    "# # get cell-type by cell matrix\n",
    "# es.max = sctype_score(scRNAseqData = scte[[\"RNA\"]]@scale.data, scaled = TRUE, \n",
    "#                       gs = gs_list$gs_positive, gs2 = gs_list$gs_negative) \n",
    "\n",
    "# # merge by cluster\n",
    "# cL_resutls = do.call(\"rbind\", lapply(unique(scte@meta.data$seurat_clusters), function(cl){\n",
    "#     es.max.cl = sort(rowSums(es.max[ ,rownames(scte@meta.data[scte@meta.data$seurat_clusters==cl, ])]), decreasing = !0)\n",
    "#     head(data.frame(cluster = cl, type = names(es.max.cl), scores = es.max.cl, ncells = sum(scte@meta.data$seurat_clusters==cl)), 10)\n",
    "# }))\n",
    "# sctype_scores = cL_resutls %>% group_by(cluster) %>% top_n(n = 1, wt = scores)  \n",
    "\n",
    "# # set low-confident (low ScType score) clusters to \"unknown\"\n",
    "# sctype_scores$type[as.numeric(as.character(sctype_scores$scores)) < sctype_scores$ncells/4] = \"Unknown\"\n",
    "\n",
    "\n",
    "# scte@meta.data$customclassif = \"\"\n",
    "# for(j in unique(sctype_scores$cluster)){\n",
    "#   cl_type = sctype_scores[sctype_scores$cluster==j,]; \n",
    "#   scte@meta.data$customclassif[scte@meta.data$seurat_clusters == j] = as.character(cl_type$type[1])\n",
    "# }\n",
    "\n",
    "# ## save cell type\n",
    "# cell_types<-scte[['customclassif']]\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
