{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_349084/3626175138.py:22: DtypeWarning: Columns (0,1,2,3,4,6,7,8,13,14,15,16) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  rmsk=pd.read_table(rmsk_f,sep='\\t',header=None)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AD_HS_00004AD_HS_00005AD_HS_00001AD_HS_00002AD_HS_00003AD_HS_00007MS_HS_00001AD_HS_00006AD_HS_00008MS_HS_00002PD_HS_00001\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "loading data: AD_HS_00004_94\n",
      "loading data: AD_HS_00005_658loading data: AD_HS_00001_279loading data: AD_HS_00002_624loading data: AD_HS_00003_336loading data: AD_HS_00007_562loading data: AD_HS_00008_571loading data: AD_HS_00006_752loading data: MS_HS_00002_166loading data: PD_HS_00001_55loading data: MS_HS_00001_612\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "['AD_HS_00001', 'AD_HS_00002', 'AD_HS_00003', 'AD_HS_00004', 'AD_HS_00005', 'AD_HS_00006', 'AD_HS_00007', 'AD_HS_00008', 'MS_HS_00001', 'MS_HS_00002', 'PD_HS_00001']\n",
      "extracting disease v.s. control\n",
      "extracting cell v.s. cell\n",
      "extracting disease v.s. control\n",
      "extracting cell v.s. cell\n",
      "extracting disease v.s. control\n",
      "extracting cell v.s. cell\n",
      "extracting disease v.s. control\n",
      "extracting cell v.s. cell\n",
      "extracting disease v.s. control\n",
      "extracting cell v.s. cell\n",
      "extracting disease v.s. control\n",
      "extracting cell v.s. cell\n",
      "extracting disease v.s. control\n",
      "extracting cell v.s. cell\n",
      "extracting disease v.s. control\n",
      "extracting cell v.s. cell\n"
     ]
    }
   ],
   "source": [
    "###\n",
    " # @author [Wankun Deng]\n",
    " # @email [dengwankun@gmail.com]\n",
    " # @create date 2023-06-30 13:17:49\n",
    " # @modify date 2023-07-07 14:09:09\n",
    " # @desc [description]\n",
    "###\n",
    "import os\n",
    "import sys\n",
    "import random\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from multiprocessing import Pool\n",
    "from scipy.stats import mannwhitneyu\n",
    "\n",
    "sys.argv=['th','../data/all_datasets','../www/mysql','../../universal_data/rmsk/rmsk_GRCh38.txt']\n",
    "input_path=sys.argv[1]\n",
    "output_path=sys.argv[2]\n",
    "rmsk_f=sys.argv[3]\n",
    "\n",
    "file_list=[x for x in os.listdir(input_path) if x.endswith('.cell_exp.txt')]\n",
    "rmsk=pd.read_table(rmsk_f,sep='\\t',header=None)\n",
    "rtes=rmsk[rmsk[11].isin(['LINE','SINE','LTR'])][10].unique()\n",
    "datasets=list(set([x.split('.')[0] for x in file_list]))\n",
    "datasets.sort()\n",
    "cell_types=['Ex','Ast','Mic','OPC','Oli','In']\n",
    "def get_dataset(dataset):\n",
    "    print(dataset)\n",
    "    randint=random.randint(0,1000)\n",
    "    print(f'loading data: {dataset}_{randint}')\n",
    "    results=[]\n",
    "    dt_ls=[x for x in file_list if x.startswith(dataset)]\n",
    "    cell_exp=pd.read_table(os.path.join(input_path,dt_ls[0]),index_col=0)\n",
    "    if len(dt_ls) >1:\n",
    "        cell_umap=pd.read_table(f'{input_path}/{dataset}.1.cell_umap.txt',index_col=0)\n",
    "        for i in range(1,len(dt_ls)):\n",
    "            cell_exp=pd.concat([cell_exp,pd.read_table(f'{input_path}/'+dt_ls[i],index_col=0)])\n",
    "            cell_umap=pd.concat([cell_umap,pd.read_table(f'{input_path}/{dataset}.{i}.cell_umap.txt',index_col=0)])\n",
    "    else:\n",
    "        cell_umap=pd.read_table(f'{input_path}/'+dataset+'.cell_umap.txt',index_col=0)\n",
    "    \n",
    "    cell_umap['predicted.celltype'] = cell_umap['predicted.celltype'].replace(\n",
    "        'Opc', 'OPC')\n",
    "    for i in range(cell_umap.shape[0]):\n",
    "        if cell_umap.iloc[i,1] =='Stage_0':\n",
    "            cell_umap.iloc[i,1]='Control'\n",
    "        if cell_umap.iloc[i,1] !='Control':\n",
    "            cell_umap.iloc[i,1]=cell_umap.iloc[i,7].split('_')[0]\n",
    "\n",
    "    cell_exp=cell_exp.loc[:,[x for x in cell_exp.columns if x in rtes]]\n",
    "    ## disease v.s. control\n",
    "    print('extracting disease v.s. control')\n",
    "    disease=dataset.split('_')[0]\n",
    "    for cell in cell_types:\n",
    "        disease_list=cell_umap.loc[(cell_umap['predicted.celltype']==cell) & (cell_umap['Diagnosis']!='Control'),:].index\n",
    "        control_list=cell_umap.loc[(cell_umap['predicted.celltype']==cell) & (cell_umap['Diagnosis']=='Control'),:].index\n",
    "        for te in rtes:\n",
    "            if te in cell_exp.columns:\n",
    "                disease_exp=np.exp(cell_exp.loc[disease_list,te])\n",
    "                control_exp=np.exp(cell_exp.loc[control_list,te])\n",
    "                pval=mannwhitneyu(disease_exp,control_exp)[1]\n",
    "                c1_median=np.mean(disease_exp)\n",
    "                c2_median=np.mean(control_exp)\n",
    "                pce1=np.count_nonzero(disease_exp>1)/(1.0*len(disease_exp))\n",
    "                pce2=np.count_nonzero(control_exp>1)/(1.0*len(control_exp))\n",
    "                fc=np.log2(np.mean(disease_exp)/np.mean(control_exp))\n",
    "                results.append([te,f'{disease} v.s. Control',dataset,cell,'NA',np.log(c1_median),np.log(c2_median),fc,pval,pce1,pce2,len(disease_exp),len(control_exp)])\n",
    "    ## cell v.s. cell\n",
    "    print('extracting cell v.s. cell')\n",
    "    for cell1 in cell_types:\n",
    "        for cell2 in cell_types:\n",
    "            if not cell1==cell2:\n",
    "                for condition in [disease,'Control']:\n",
    "                    cell1_list=cell_umap.loc[(cell_umap['predicted.celltype']==cell1) & (cell_umap['Diagnosis']==condition),:].index\n",
    "                    cell2_list=cell_umap.loc[(cell_umap['predicted.celltype']==cell2) & (cell_umap['Diagnosis']==condition),:].index\n",
    "                    for te in rtes:\n",
    "                        if te in cell_exp.columns:\n",
    "                            cell1_exp=np.exp(cell_exp.loc[cell1_list,te])\n",
    "                            cell2_exp=np.exp(cell_exp.loc[cell2_list,te])\n",
    "                            pval=mannwhitneyu(cell1_exp,cell2_exp)[1]\n",
    "                            c1_median=np.mean(cell1_exp)\n",
    "                            c2_median=np.mean(cell2_exp)\n",
    "                            fc=np.log2(np.mean(cell1_exp)/np.mean(cell2_exp))\n",
    "                            pce1=np.count_nonzero(cell1_exp>1)/(1.0*len(cell1_exp))\n",
    "                            pce2=np.count_nonzero(cell2_exp>1)/(1.0*len(cell2_exp))\n",
    "                            results.append([te,f'{cell1} v.s. {cell2}',dataset,'NA',condition,np.log(c1_median),np.log(c2_median),fc,pval,pce1,pce2,len(cell1_exp),len(cell2_exp)])\n",
    "    return results\n",
    "\n",
    "pool = Pool(20)\n",
    "print(datasets)\n",
    "ret=pool.map(get_dataset,datasets)\n",
    "pool.close()\n",
    "pool.join()\n",
    "\n",
    "final_result=[]\n",
    "for i in ret:\n",
    "    final_result.extend(i)\n",
    "final_result=pd.DataFrame(final_result,columns=['TE','Comparison','Dataset','Cell','Condition','log2_median1','log2_median2','log2_FC','pval','pce1','pce2','cell1_count','cell2_count'])\n",
    "\n",
    "te_de=open(f'{output_path}/exp_de.sql','w')\n",
    "\n",
    "te_de.write('''use scARE;\n",
    "    DROP TABLE IF EXISTS `EXP_DE`;\n",
    "    CREATE TABLE `EXP_DE` (\n",
    "        `TE` varchar(255) NOT NULL,\n",
    "        `COMPARISON` varchar(255) NOT NULL,\n",
    "        `DATASET` varchar(255) NOT NULL,\n",
    "        `CELL` varchar(255) NOT NULL,\n",
    "        `CONDITION` varchar(255) NOT NULL,\n",
    "        `MEAN1` float NOT NULL,\n",
    "        `MEAN2` float NOT NULL,\n",
    "        `FC` float NOT NULL,\n",
    "        `PVAL` float NOT NULL,\n",
    "        `PCE1` float NOT NULL,\n",
    "        `PCE2` float NOT NULL,\n",
    "        `CELL1_COUNT` int NOT NULL,\n",
    "        `CELL2_COUNT` int NOT NULL);\n",
    "        set autocommit=0;\\n''')\n",
    "\n",
    "count=0\n",
    "values_list=[]\n",
    "template='INSERT INTO `EXP_DE` (TE,COMPARISON,DATASET,CELL,`CONDITION`,MEAN1,MEAN2,`FC`,PVAL,PCE1,PCE2,CELL1_COUNT,CELL2_COUNT) VALUES {values};\\n'\n",
    "for i in range(final_result.shape[0]):\n",
    "    if count==2000:\n",
    "        te_de.write(template.format(values=','.join(values_list)))\n",
    "        count=0\n",
    "        values_list=[]\n",
    "    row=list(final_result.iloc[i,:])\n",
    "    row[:5]=['\"'+str(x)+'\"' for x in row[:5]]\n",
    "    values_list.append('('+','.join([str(x) for x in row])+')')\n",
    "    count+=1\n",
    "\n",
    "if count>0:\n",
    "    te_de.write(template.format(values=','.join(values_list)))\n",
    "te_de.write('commit;')\n",
    "te_de.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## assign cell type according to ScType tuturial\n",
    "\n",
    "## load sc type and database\n",
    "# source(paste0(sctype_path,\"/R/gene_sets_prepare.R\"))\n",
    "# source(paste0(sctype_path,\"/R/sctype_score_.R\"))\n",
    "# gs_list = gene_sets_prepare(paste0(sctype_path,'/ScTypeDB_full.xlsx'), \"Brain\" )\n",
    "# # get cell-type by cell matrix\n",
    "# es.max = sctype_score(scRNAseqData = scte[[\"RNA\"]]@scale.data, scaled = TRUE, \n",
    "#                       gs = gs_list$gs_positive, gs2 = gs_list$gs_negative) \n",
    "\n",
    "# # merge by cluster\n",
    "# cL_resutls = do.call(\"rbind\", lapply(unique(scte@meta.data$seurat_clusters), function(cl){\n",
    "#     es.max.cl = sort(rowSums(es.max[ ,rownames(scte@meta.data[scte@meta.data$seurat_clusters==cl, ])]), decreasing = !0)\n",
    "#     head(data.frame(cluster = cl, type = names(es.max.cl), scores = es.max.cl, ncells = sum(scte@meta.data$seurat_clusters==cl)), 10)\n",
    "# }))\n",
    "# sctype_scores = cL_resutls %>% group_by(cluster) %>% top_n(n = 1, wt = scores)  \n",
    "\n",
    "# # set low-confident (low ScType score) clusters to \"unknown\"\n",
    "# sctype_scores$type[as.numeric(as.character(sctype_scores$scores)) < sctype_scores$ncells/4] = \"Unknown\"\n",
    "\n",
    "\n",
    "# scte@meta.data$customclassif = \"\"\n",
    "# for(j in unique(sctype_scores$cluster)){\n",
    "#   cl_type = sctype_scores[sctype_scores$cluster==j,]; \n",
    "#   scte@meta.data$customclassif[scte@meta.data$seurat_clusters == j] = as.character(cl_type$type[1])\n",
    "# }\n",
    "\n",
    "# ## save cell type\n",
    "# cell_types<-scte[['customclassif']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "library(stringr)\n",
    "\n",
    "lapply(c(\"dplyr\",\"Seurat\",\"HGNChelper\",\"openxlsx\"), library, character.only = T)\n",
    "\n",
    "load('../data/reference.RData')\n",
    "\n",
    "sample_meta<-'AD_HS_00001\tMolecular characterization of selectively vulnerable neurons in Alzheimer’s Disease\t20\tH.Sapiens\tAD\tSuperior frontal gyrus;Superior frontal gyrus;Superior frontal gyrus;Superior frontal gyrus;Superior frontal gyrus;Superior frontal gyrus;Superior frontal gyrus;Superior frontal gyrus;Superior frontal gyrus;Superior frontal gyrus;Entorhinal cortex;Entorhinal cortex;Entorhinal cortex;Entorhinal cortex;Entorhinal cortex;Entorhinal cortex;Entorhinal cortex;Entorhinal cortex;Entorhinal cortex;Entorhinal cortex\tBraak:0;0;0;2;2;2;2;6;6;6;0;0;0;2;2;2;2;6;6;6\tExcitatoryNeuron;InhibitoryNeuron;Oligodendrocyte;OPC;Astrocyte;Microglia;Endothelia;Pericyte\tGSE147528\tSingle-nucleus RNA-seq\t10x Genomics\tMale\t60;50;71;72;87;91;77;82;72;82;60;50;71;72;87;91;77;82;72;82\tSRR11422700;SRR11422701;SRR11422702;SRR11422703;SRR11422704;SRR11422705;SRR11422706;SRR11422707;SRR11422708;SRR11422709;SRR11422710;SRR11422711;SRR11422712;SRR11422713;SRR11422714;SRR11422715;SRR11422716;SRR11422717;SRR11422718;SRR11422719'\n",
    "split_meta<-unlist(strsplit(sample_meta, '\\t'))\n",
    "samples<-unlist(strsplit(split_meta[14],';'))\n",
    "disease<-rep(split_meta[5],length(samples))\n",
    "stage<-unlist(strsplit(unlist(strsplit(split_meta[7],':'))[2],';'))\n",
    "gender<-rep(split_meta[12],length(samples))\n",
    "age<-unlist(strsplit(split_meta[13],';'))\n",
    "\n",
    "sample_meta<-data.frame(samples=samples,disease=disease,stage=stage,gender=gender,age=age)\n",
    "sample_meta[sample_meta$stage=='0','disease']<-'Control'\n",
    "rownames(sample_meta)<-sample_meta$samples\n",
    "\n",
    "# args <- commandArgs(trailingOnly=TRUE)\n",
    "args<-c('../data/3/scte','../data/3/cell_umap.txt','SRR11422700','/home/wdeng3/workspace/software/sc-type','../../universal_data/rmsk/rmsk_GRCh38.txt')\n",
    "data_path<-args[1]\n",
    "out_path<-args[2]\n",
    "sample_tag<-args[3]\n",
    "sctype_path<-args[4]\n",
    "rmsk_path<-args[5]\n",
    "rmsk<-read.table(rmsk_path,sep='\\t',header=F)\n",
    "to_remove<-unlist(rmsk[!(rmsk$V12 %in% c('LINE','SINE','LTR')),'V11'])\n",
    "to_remove<-str_replace_all(to_remove,'-','\\\\.')\n",
    "\n",
    "reps<-unlist(rmsk[rmsk$V12 %in% c('LINE','SINE','LTR'),'V11'])\n",
    "reps<-str_replace_all(reps,'-','\\\\.')\n",
    "\n",
    "create_seurat<-function(sample_tag){\n",
    "    ## load data\n",
    "    scte.data <- t(read.csv(paste0(data_path,'/',sample_tag,'/',sample_tag,'.csv'),check.names=F, row.names = 1))\n",
    "    # remove repeats from gene expression matrix\n",
    "\n",
    "    scte.gene<-scte.data[!(row.names(scte.data) %in% to_remove),]\n",
    "    # normalize and scale data\n",
    "    scte <- CreateSeuratObject(counts = scte.gene, project = sample_tag, min.cells = 3, min.features = 200)\n",
    "    # normalize data\n",
    "    scte[[\"percent.mt\"]] <- PercentageFeatureSet(scte, pattern = \"^MT.|^MT-\")\n",
    "    scte <- subset(scte, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)\n",
    "    scte <- SCTransform(scte, vars.to.regress = \"percent.mt\", verbose = FALSE)\n",
    "\n",
    "    cell_meta<-scte@meta.data\n",
    "    cell_meta$disease<-sample_meta[cell_meta$orig.ident,'disease']\n",
    "\n",
    "    cell_meta$stage<-sample_meta[cell_meta$orig.ident,'stage']\n",
    "    cell_meta$gender<-sample_meta[cell_meta$orig.ident,'gender']\n",
    "    cell_meta$age<-sample_meta[cell_meta$orig.ident,'age']\n",
    "    scte@meta.data<-cell_meta\n",
    "    scte\n",
    "}\n",
    "\n",
    "scte<-create_seurat(sample_tag)\n",
    "\n",
    "scte <- FindVariableFeatures(scte, selection.method = \"vst\", nfeatures = 2000)\n",
    "# scale and run PCA\n",
    "scte <- ScaleData(scte, features = rownames(scte))\n",
    "scte <- RunPCA(scte, features = VariableFeatures(object = scte))\n",
    "\n",
    "sdv<-Stdev(scte,reduction='pca')\n",
    "sdv<-sdv[sdv>2]\n",
    "npca<-length(sdv)\n",
    "\n",
    "scte <- FindNeighbors(scte, dims = 1:npca)\n",
    "scte <- FindClusters(scte, resolution = 0.8)\n",
    "scte <- RunUMAP(scte, dims = 1:npca)\n",
    "\n",
    "\n",
    "scte.anchors <- FindTransferAnchors(reference = reference, query = scte,\n",
    "    dims = 1:30, reference.reduction = \"pca\")\n",
    "scte.query <- MapQuery(anchorset = scte.anchors, reference = reference, query = scte,\n",
    "    refdata = list(celltype = \"broad.cell.type\"), reference.reduction = \"pca\", reduction.model = \"umap\")\n",
    "cell_meta<-scte.query@meta.data\n",
    "\n",
    "cell_meta$UMAP_1<-scte.query@reductions[['umap']]@cell.embeddings[,1]\n",
    "cell_meta$UMAP_2<-scte.query@reductions[['umap']]@cell.embeddings[,2]\n",
    "scte.query@meta.data<-cell_meta\n",
    "write.table(cell_meta,out_path,sep='\\t',quote=F,row.names=T)\n",
    "\n",
    "## get gene expression from scte_query\n",
    "scte.query.gene<-as.data.frame(t(as.data.frame(GetAssayData(scte.query, slot = \"data\"))))\n",
    "scte.query.gene$UMAP_1<-scte.query@meta.data$UMAP_1\n",
    "scte.query.gene$UMAP_2<-scte.query@meta.data$UMAP_2\n",
    "write.table(scte.query.gene,'../data/3/cell_exp.txt',sep='\\t',quote=F,row.names=T)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
