{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "library(stringr)\n",
    "\n",
    "lapply(c(\"dplyr\",\"Seurat\",\"HGNChelper\",\"openxlsx\"), library, character.only = T)\n",
    "\n",
    "load('../data/reference.RData')\n",
    "\n",
    "sample_meta<-'AD_HS_00001\tMolecular characterization of selectively vulnerable neurons in Alzheimerâ€™s Disease\t20\tH.Sapiens\tAD\tSuperior frontal gyrus;Superior frontal gyrus;Superior frontal gyrus;Superior frontal gyrus;Superior frontal gyrus;Superior frontal gyrus;Superior frontal gyrus;Superior frontal gyrus;Superior frontal gyrus;Superior frontal gyrus;Entorhinal cortex;Entorhinal cortex;Entorhinal cortex;Entorhinal cortex;Entorhinal cortex;Entorhinal cortex;Entorhinal cortex;Entorhinal cortex;Entorhinal cortex;Entorhinal cortex\tBraak:0;0;0;2;2;2;2;6;6;6;0;0;0;2;2;2;2;6;6;6\tExcitatoryNeuron;InhibitoryNeuron;Oligodendrocyte;OPC;Astrocyte;Microglia;Endothelia;Pericyte\tGSE147528\tSingle-nucleus RNA-seq\t10x Genomics\tMale\t60;50;71;72;87;91;77;82;72;82;60;50;71;72;87;91;77;82;72;82\tSRR11422700;SRR11422701;SRR11422702;SRR11422703;SRR11422704;SRR11422705;SRR11422706;SRR11422707;SRR11422708;SRR11422709;SRR11422710;SRR11422711;SRR11422712;SRR11422713;SRR11422714;SRR11422715;SRR11422716;SRR11422717;SRR11422718;SRR11422719'\n",
    "split_meta<-unlist(strsplit(sample_meta, '\\t'))\n",
    "samples<-unlist(strsplit(split_meta[14],';'))\n",
    "disease<-rep(split_meta[5],length(samples))\n",
    "stage<-unlist(strsplit(unlist(strsplit(split_meta[7],':'))[2],';'))\n",
    "gender<-rep(split_meta[12],length(samples))\n",
    "age<-unlist(strsplit(split_meta[13],';'))\n",
    "\n",
    "sample_meta<-data.frame(samples=samples,disease=disease,stage=stage,gender=gender,age=age)\n",
    "sample_meta[sample_meta$stage=='0','disease']<-'Control'\n",
    "rownames(sample_meta)<-sample_meta$samples\n",
    "\n",
    "# args <- commandArgs(trailingOnly=TRUE)\n",
    "args<-c('../data/3/scte','../data/3/cell_umap.txt','SRR11422700','/home/wdeng3/workspace/software/sc-type','../../universal_data/rmsk/rmsk_GRCh38.txt')\n",
    "data_path<-args[1]\n",
    "out_path<-args[2]\n",
    "sample_tag<-args[3]\n",
    "sctype_path<-args[4]\n",
    "rmsk_path<-args[5]\n",
    "rmsk<-read.table(rmsk_path,sep='\\t',header=F)\n",
    "to_remove<-unlist(rmsk[!(rmsk$V12 %in% c('LINE','SINE','LTR')),'V11'])\n",
    "to_remove<-str_replace_all(to_remove,'-','\\\\.')\n",
    "\n",
    "reps<-unlist(rmsk[rmsk$V12 %in% c('LINE','SINE','LTR'),'V11'])\n",
    "reps<-str_replace_all(reps,'-','\\\\.')\n",
    "\n",
    "create_seurat<-function(sample_tag){\n",
    "    ## load data\n",
    "    scte.data <- t(read.csv(paste0(data_path,'/',sample_tag,'/',sample_tag,'.csv'),check.names=F, row.names = 1))\n",
    "    # remove repeats from gene expression matrix\n",
    "\n",
    "    scte.gene<-scte.data[!(row.names(scte.data) %in% to_remove),]\n",
    "    # normalize and scale data\n",
    "    scte <- CreateSeuratObject(counts = scte.gene, project = sample_tag, min.cells = 3, min.features = 200)\n",
    "    # normalize data\n",
    "    scte[[\"percent.mt\"]] <- PercentageFeatureSet(scte, pattern = \"^MT.|^MT-\")\n",
    "    scte <- subset(scte, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)\n",
    "    scte <- SCTransform(scte, vars.to.regress = \"percent.mt\", verbose = FALSE)\n",
    "\n",
    "    cell_meta<-scte@meta.data\n",
    "    cell_meta$disease<-sample_meta[cell_meta$orig.ident,'disease']\n",
    "\n",
    "    cell_meta$stage<-sample_meta[cell_meta$orig.ident,'stage']\n",
    "    cell_meta$gender<-sample_meta[cell_meta$orig.ident,'gender']\n",
    "    cell_meta$age<-sample_meta[cell_meta$orig.ident,'age']\n",
    "    scte@meta.data<-cell_meta\n",
    "    scte\n",
    "}\n",
    "\n",
    "scte<-create_seurat(sample_tag)\n",
    "\n",
    "scte <- FindVariableFeatures(scte, selection.method = \"vst\", nfeatures = 2000)\n",
    "# scale and run PCA\n",
    "scte <- ScaleData(scte, features = rownames(scte))\n",
    "scte <- RunPCA(scte, features = VariableFeatures(object = scte))\n",
    "\n",
    "sdv<-Stdev(scte,reduction='pca')\n",
    "sdv<-sdv[sdv>2]\n",
    "npca<-length(sdv)\n",
    "\n",
    "scte <- FindNeighbors(scte, dims = 1:npca)\n",
    "scte <- FindClusters(scte, resolution = 0.8)\n",
    "scte <- RunUMAP(scte, dims = 1:npca)\n",
    "\n",
    "\n",
    "scte.anchors <- FindTransferAnchors(reference = reference, query = scte,\n",
    "    dims = 1:30, reference.reduction = \"pca\")\n",
    "scte.query <- MapQuery(anchorset = scte.anchors, reference = reference, query = scte,\n",
    "    refdata = list(celltype = \"broad.cell.type\"), reference.reduction = \"pca\", reduction.model = \"umap\")\n",
    "cell_meta<-scte.query@meta.data\n",
    "\n",
    "cell_meta$UMAP_1<-scte.query@reductions[['umap']]@cell.embeddings[,1]\n",
    "cell_meta$UMAP_2<-scte.query@reductions[['umap']]@cell.embeddings[,2]\n",
    "scte.query@meta.data<-cell_meta\n",
    "write.table(cell_meta,out_path,sep='\\t',quote=F,row.names=T)\n",
    "\n",
    "## get gene expression from scte_query\n",
    "scte.query.gene<-as.data.frame(t(as.data.frame(GetAssayData(scte.query, slot = \"data\"))))\n",
    "scte.query.gene$UMAP_1<-scte.query@meta.data$UMAP_1\n",
    "scte.query.gene$UMAP_2<-scte.query@meta.data$UMAP_2\n",
    "write.table(scte.query.gene,'../data/3/cell_exp.txt',sep='\\t',quote=F,row.names=T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_9533/1952009851.py:25: DtypeWarning: Columns (0,1,2,3,4,6,7,8,13,14,15,16) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  rmsk=pd.read_csv(rmsk_f,sep='\\t',header=None)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "import sys\n",
    "import math\n",
    "import tqdm\n",
    "import pandas as pd\n",
    "import os\n",
    "import json\n",
    "from collections import defaultdict\n",
    "\n",
    "sys.argv=['th','../data/all_datasets','../www/mysql','overwrite','../../universal_data/ref/GRCh38/gencode.v43.basic.annotation.gtf','../../universal_data/rmsk/rmsk_GRCh38.txt']\n",
    "inpath=sys.argv[1]\n",
    "out_path=sys.argv[2]\n",
    "mode='append' if sys.argv[3]=='append' else 'overwrite'\n",
    "gtf=sys.argv[4]\n",
    "rmsk_f=sys.argv[5]\n",
    "\n",
    "if mode=='overwrite':\n",
    "    genes=[]\n",
    "    for line in open(gtf):\n",
    "        if not line.startswith('#'):\n",
    "            info=line.strip().split('\\t')\n",
    "            if info[2]=='gene' and ('lncRNA' in info[-1] or 'protein_coding' in info[-1]):\n",
    "                split_='gene_name \"' if 'gene_name' in info[-1] else 'gene_id \"'\n",
    "                genes.append(info[-1].split(split_)[1].split('\"')[0])            \n",
    "\n",
    "    rmsk=pd.read_csv(rmsk_f,sep='\\t',header=None)\n",
    "    rmsk=rmsk[rmsk[11].isin(['LINE','SINE','LTR'])]\n",
    "    genes.extend(rmsk[10].unique())\n",
    "    genes=list(set(genes))\n",
    "\n",
    "    exp_tables=[open(f'{out_path}/cell_exp_{i}.sql','w') for i in range(math.ceil((len(genes))/1000.0))]\n",
    "    for i in range(len(exp_tables)):\n",
    "        exp_tables[i].write('''CREATE DATABASE IF NOT EXISTS scARE;\n",
    "                                USE scARE;\n",
    "                                DROP TABLE IF EXISTS CELL_EXP_{i}; \n",
    "                                CREATE TABLE CELL_EXP_{i} (\n",
    "                                scARE_ID varchar(255) NOT NULL,\n",
    "                                CELL varchar(255) NOT NULL,\\n'''.format(i=i))\n",
    "        \n",
    "    gene_dict=open(f'{out_path}/gene_dict.sql','w') \n",
    "    gene_dict.write('''\\\n",
    "            CREATE DATABASE IF NOT EXISTS scARE;\n",
    "            USE scARE;\n",
    "            DROP TABLE IF EXISTS GENE_DICT;\n",
    "            CREATE TABLE GENE_DICT (\n",
    "                GENE varchar(255) NOT NULL,\n",
    "                TABLE_ID INT NOT NULL);\\n''')\n",
    "\n",
    "    table_genes=defaultdict(list)\n",
    "    for i in range(len(genes)):\n",
    "        index=math.floor(i/1000.0)\n",
    "        exp_tables[index].write(f'`{genes[i]}` FLOAT DEFAULT 0,\\n')\n",
    "        table_genes[index].append(genes[i])\n",
    "        gene_dict.write(f'INSERT INTO GENE_DICT values(\"{genes[i]}\",\"{index}\");\\n')\n",
    "    for i in range(len(exp_tables)):\n",
    "        exp_tables[i].write('UMAP_1 FLOAT NOT NULL,\\n')\n",
    "        exp_tables[i].write('UMAP_2 FLOAT NOT NULL);\\n')\n",
    "        exp_tables[i].write('set autocommit=0;\\n')\n",
    "        exp_tables[i].flush()\n",
    "    json.dump(table_genes,open(f'{out_path}/table_genes.json','w'))\n",
    "else:\n",
    "    table_genes=json.load(open(f'{out_path}/table_genes.json','r'))\n",
    "    exp_tables=[open(f'{out_path}/cell_exp_{i}.sql','a') for i in range(len(table_genes))]\n",
    "\n",
    "\n",
    "def one_table(i):\n",
    "    expressed_i=[x for x in cell_exp.columns if x in table_genes[i]]+['UMAP_1','UMAP_2']\n",
    "    tmp_exp=pd.DataFrame(columns=(['scARE_ID','CELL']+expressed_i),index=cell_exp.index)\n",
    "    tmp_exp['scARE_ID']=['\"%s\"'%cell_dataset[x] for x in cell_exp.index]\n",
    "    tmp_exp['CELL']=['\"%s\"'%x for x in cell_exp.index]\n",
    "    tmp_exp.loc[:,expressed_i]=cell_exp.loc[:,expressed_i]\n",
    "    all_values=tmp_exp.agg(lambda x:  ','.join(x.astype(str)), axis=1).values\n",
    "    expressed_i_c=','.join(expressed_i)\n",
    "    template=f'''INSERT INTO CELL_EXP_{i} (scARE_ID,CELL,{expressed_i_c}) values '''\n",
    "    count=0\n",
    "    values_list=[]\n",
    "    to_write=''\n",
    "    for x in all_values:\n",
    "        if count==2000:\n",
    "            to_write+=template+','.join(values_list)+';\\n'\n",
    "            count=0\n",
    "            values_list=[]\n",
    "        values_list.append('('+x+')')\n",
    "        count+=1\n",
    "    if len(values_list)>0:\n",
    "        to_write+=template+','.join(values_list)+';\\n'\n",
    "    # to_write='\\n'.join([template.format(i=i,expressed_i_c=expressed_i_c,values=x) for x in all_values])+'\\n'\n",
    "    exp_tables[i].write(to_write)\n",
    "\n",
    "from multiprocessing import Pool\n",
    "for file_ in os.listdir(inpath):\n",
    "    if file_.endswith('.cell_exp.txt'):\n",
    "        cell_exp=pd.read_csv(os.path.join(inpath,file_),sep='\\t',index_col=0)\n",
    "        cell_dataset=pd.read_csv(os.path.join(inpath,file_.replace('cell_exp','cell_umap')),sep='\\t',index_col=0)['dataset'].to_dict()\n",
    "        pool=Pool(40)\n",
    "        pool.map(one_table,range(len(exp_tables)))\n",
    "        pool.close()\n",
    "        pool.join()\n",
    "\n",
    "[x.write('commit;\\n') for x in exp_tables]\n",
    "[x.flush() for x in exp_tables]\n",
    "[x.close() for x in exp_tables]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## assign cell type according to ScType tuturial\n",
    "\n",
    "## load sc type and database\n",
    "# source(paste0(sctype_path,\"/R/gene_sets_prepare.R\"))\n",
    "# source(paste0(sctype_path,\"/R/sctype_score_.R\"))\n",
    "# gs_list = gene_sets_prepare(paste0(sctype_path,'/ScTypeDB_full.xlsx'), \"Brain\" )\n",
    "# # get cell-type by cell matrix\n",
    "# es.max = sctype_score(scRNAseqData = scte[[\"RNA\"]]@scale.data, scaled = TRUE, \n",
    "#                       gs = gs_list$gs_positive, gs2 = gs_list$gs_negative) \n",
    "\n",
    "# # merge by cluster\n",
    "# cL_resutls = do.call(\"rbind\", lapply(unique(scte@meta.data$seurat_clusters), function(cl){\n",
    "#     es.max.cl = sort(rowSums(es.max[ ,rownames(scte@meta.data[scte@meta.data$seurat_clusters==cl, ])]), decreasing = !0)\n",
    "#     head(data.frame(cluster = cl, type = names(es.max.cl), scores = es.max.cl, ncells = sum(scte@meta.data$seurat_clusters==cl)), 10)\n",
    "# }))\n",
    "# sctype_scores = cL_resutls %>% group_by(cluster) %>% top_n(n = 1, wt = scores)  \n",
    "\n",
    "# # set low-confident (low ScType score) clusters to \"unknown\"\n",
    "# sctype_scores$type[as.numeric(as.character(sctype_scores$scores)) < sctype_scores$ncells/4] = \"Unknown\"\n",
    "\n",
    "\n",
    "# scte@meta.data$customclassif = \"\"\n",
    "# for(j in unique(sctype_scores$cluster)){\n",
    "#   cl_type = sctype_scores[sctype_scores$cluster==j,]; \n",
    "#   scte@meta.data$customclassif[scte@meta.data$seurat_clusters == j] = as.character(cl_type$type[1])\n",
    "# }\n",
    "\n",
    "# ## save cell type\n",
    "# cell_types<-scte[['customclassif']]\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "916dbcbb3f70747c44a77c7bcd40155683ae19c65e1c03b4aa3499c5328201f1"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
